### PR TITLE
add the ability to create a bpf program without device handle

### PR DESCRIPTION
**Pull request by sensokame**
I think it would be a good addition to add the ability to create BpfPrograms without the need of device handles. (create general purpose bpfprogram)
So I added a create / trycreate methods that use linktype and pcap_open_dead instead of device handles.

(I needed this feature in my work)

### CODE REVIEW

**Comment by sensokame** [comment]
@kayoub5 can you check this branch and merge it if it's useful

File: SharpPcap/LibPcap/BpfProgram.cs [hunk 1]
```diff
@@ -82,6 +82,26 @@ public static BpfProgram Create(PcapHandle pcapHandle, string filter, int optimi
             return bpfProgram;
         }
 
+        public static BpfProgram TryCreate(int linktype, string filter, int optimize = 1, uint netmask = 0)
```
Comment (kayoub5) [inline SharpPcap/LibPcap/BpfProgram.cs:85]:
```suggestion
        public static BpfProgram TryCreate(LinkLayers linktype, string filter, int optimize = 1, uint netmask = 0)
```

File: SharpPcap/LibPcap/BpfProgram.cs [hunk 2]
```diff
@@ -82,6 +82,26 @@ public static BpfProgram Create(PcapHandle pcapHandle, string filter, int optimi
...
+        {
+            using (var handle = LibPcapSafeNativeMethods.pcap_open_dead(linktype, 65535))
```
Comment (kayoub5) [inline SharpPcap/LibPcap/BpfProgram.cs:87]:
```suggestion
            using (var handle = LibPcapSafeNativeMethods.pcap_open_dead(linktype, Pcap.MAX_PACKET_SIZE))
```

File: SharpPcap/LibPcap/BpfProgram.cs [hunk 3]
```diff
@@ -82,6 +82,26 @@ public static BpfProgram Create(PcapHandle pcapHandle, string filter, int optimi
...
+
+        }
+
+
+        public static BpfProgram Create(int linktype, string filter, int optimize = 1, uint netmask = 0)
+        {
+            var bpfProgram = TryCreate(linktype, filter, optimize, netmask);
+            if (bpfProgram == null)
+            {
+                throw new PcapException("Could not create bpf program using this linktype");
```
Comment (kayoub5) [inline SharpPcap/LibPcap/BpfProgram.cs:100]:
```suggestion
                throw new PcapException(PcapDevice.GetLastError(pcapHandle));
```

**Comment by sensokame** [comment]
@kayoub5 I made the changes, is it good now?

**Review by kayoub5 — APPROVED** [review]

**Review by kayoub5 — CHANGES_REQUESTED** [review]
Unit Tests

**Comment by sensokame** [comment]
@kayoub5 added unit test, can you check now?

File: Test/CheckFilterTest.cs [hunk 4]
```diff
@@ -26,6 +27,21 @@ public void TestFilters()
         /// </summary>
         [Test]
         public void BpfProgramMatches()
+        {
+            using var device = new CaptureFileReaderDevice(TestHelper.GetFile("arp_with_vlan.pcap"));
+            device.Open();
+
+            using var bpfProgram = BpfProgram.Create(LinkLayers.Ethernet, "((dst host 192.168.42.1) and (arp or icmp6 or tcp dst port 40499)) or ((vlan) and ((dst host 192.168.42.1) and (arp or icmp6 or tcp dst port 40499)))");
```
Comment (kayoub5) [inline Test/CheckFilterTest.cs:34]:
```suggestion
            var f = "(dst host 192.168.42.1) and (arp or tcp dst port 40499)";
            // Make filter work with or without VLAN
            using var bpfProgram = BpfProgram.Create(LinkLayers.Ethernet, $"({f}) or (vlan and ({f})");
```

**Comment by sensokame** [comment]
@kayoub5 done, all good?

**Review by kayoub5 — APPROVED** [review]

File: Test/CheckFilterTest.cs [hunk 5]
```diff
@@ -26,6 +27,23 @@ public void TestFilters()
         /// </summary>
         [Test]
         public void BpfProgramMatches()
+        {
+            using var device = new CaptureFileReaderDevice(TestHelper.GetFile("arp_with_vlan.pcap"));
+            device.Open();
+
+            var f = "(dst host 192.168.42.1) and (arp or tcp dst port 40499)";
+            // Make filter work with or without VLAN
+            using var bpfProgram = BpfProgram.Create(LinkLayers.Ethernet, $"({f}) or (vlan and ({f})");
```
Comment (kayoub5) [inline Test/CheckFilterTest.cs:36]:
```suggestion
            using var bpfProgram = BpfProgram.Create(LinkLayers.Ethernet, $"({f}) or (vlan and ({f}))");
```

File: SharpPcap/LibPcap/BpfProgram.cs [hunk 6]
```diff
@@ -82,6 +83,24 @@ public static BpfProgram Create(PcapHandle pcapHandle, string filter, int optimi
             return bpfProgram;
         }
 
+        public static BpfProgram TryCreate(LinkLayers linktype, string filter, int optimize = 1, uint netmask = 0)
+        {
+            using (var handle = LibPcapSafeNativeMethods.pcap_open_dead((int)linktype, Pcap.MAX_PACKET_SIZE))
+            {
+                return TryCreate(handle, filter, optimize, netmask);
+            }
+
```
Comment (kayoub5) [inline SharpPcap/LibPcap/BpfProgram.cs:92]:
```suggestion
```

**Pull request merged**